site/data/docs/custom-authentication.mdx

```diff
@@ -62,23 +62,77 @@ const authenticationAdapter = createAuthenticationAdapter({
 Assuming your application is already managing the authentication lifecycle in some way, you can pass the current authentication status along with your custom adapter to `RainbowKitAuthenticationProvider`, wrapping your existing `RainbowKitProvider`.
 
 ```tsx
+import { useEffect, useState } from "react";
 import {
-  createAuthenticationAdapter,
-  RainbowKitAuthenticationProvider,
+  getDefaultWallets,
   RainbowKitProvider,
-} from '@rainbow-me/rainbowkit';
+  RainbowKitAuthenticationProvider,
+  createAuthenticationAdapter,
+} from "@rainbow-me/rainbowkit";
+import { SiweMessage } from "siwe";
 import { AppProps } from 'next/app';
 import { WagmiConfig } from 'wagmi';
 
-const authenticationAdapter = createAuthenticationAdapter({
-  /* See above... */
-});
-
 export default function App({ Component, pageProps }: AppProps) {
-  // You'll need to resolve AUTHENTICATION_STATUS here
-  // using your application's authentication system.
-  // It needs to be either 'loading' (during initial load),
-  // 'unathenticated' or 'authenticated'.
+  const [authenticationStatus, setAuthenticationStatus] = useState<
+    "loading" | "unauthenticated" | "authenticated"
+  >("loading");
+  const [hasMounted, setHasMounted] = useState(false);
+
+  const authenticationAdapter = createAuthenticationAdapter({
+    getNonce: async () => {
+      const response = await fetch("/api/auth/nonce");
+      const res = await response.text();
+      return res;
+    },
+    createMessage: ({ nonce, address, chainId }) => {
+      return new SiweMessage({
+        domain: window.location.host,
+        address,
+        statement: "Sign in with Ethereum to the app.",
+        uri: window.location.origin,
+        version: "1",
+        chainId,
+        nonce,
+      });
+    },
+    getMessageBody: ({ message }) => {
+      return message.prepareMessage();
+    },
+    verify: async ({ message, signature }) => {
+      const verifyRes = await fetch("/api/auth/verify", {
+        method: "POST",
+        headers: { "Content-Type": "application/json" },
+        body: JSON.stringify({
+          message,
+          signature,
+        }),
+      });
+      setAuthenticationStatus(
+        verifyRes.ok ? "authenticated" : "unauthenticated"
+      );
+      return Boolean(verifyRes.ok);
+    },
+    signOut: async () => {
+      await fetch("/api/auth/logout");
+      setAuthenticationStatus("unauthenticated");
+    },
+  });
```

Comment:
```suggestion
  const authenticationAdapter = useMemo(() => createAuthenticationAdapter({
    getNonce: async () => {
      const response = await fetch("/api/auth/nonce");
      const res = await response.text();
      return res;
    },
    createMessage: ({ nonce, address, chainId }) => {
      return new SiweMessage({
        domain: window.location.host,
        address,
        statement: "Sign in with Ethereum to the app.",
        uri: window.location.origin,
        version: "1",
        chainId,
        nonce,
      });
    },
    getMessageBody: ({ message }) => {
      return message.prepareMessage();
    },
    verify: async ({ message, signature }) => {
      const verifyRes = await fetch("/api/auth/verify", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          message,
          signature,
        }),
      });
      setAuthenticationStatus(
        verifyRes.ok ? "authenticated" : "unauthenticated"
      );
      return Boolean(verifyRes.ok);
    },
    signOut: async () => {
      await fetch("/api/auth/logout");
      setAuthenticationStatus("unauthenticated");
    },
  }),
  []
 );
```

Shouldn't we use `useMemo` here?

Reply:
@smakosh that is interesting take as well.. is there a way for you to share how it looks like?

Reply:
Hey sorry for getting back to you late, yeah I can share a boilerplate doing that this week